dspace-api/src/main/java/org/dspace/uclouvain/core/mail/MailMetadataParserServiceImpl.java

```diff
@@ -0,0 +1,191 @@
+import org.dspace.content.authority.service.ChoiceAuthorityService;
+import org.dspace.content.service.ItemService;
+import org.dspace.core.Context;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * Service to parse some metadata of an item into something that can be used in a mail.
+ * This service is configured using instances of {@link MailMetadataParser} in 'mail-metadata-parser.xml'.
+ * 
+ * @author Michaël Pourbaix <michael.pourbaix@uclouvain.be>
+ */
+public class MailMetadataParserServiceImpl implements MailMetadataParserService {
+
+    private final Logger logger = LogManager.getLogger(MailMetadataParserServiceImpl.class);
+
+    private List<MailMetadataParser> parsers;
+
+    @Autowired
+    ChoiceAuthorityService choiceAuthorityService;
+    @Autowired
+    ItemService itemService;
+
+    /**
+     * Parse a given metadata for the given item into a pair.
+     * The pair contains the label of the field and the actual field value(s).
+     * 
+     * @param context The current DSpace context.
+     * @param item The item to extract metadata from.
+     * @param metadataField The metadata field to extract from the item.
+     * @param lang The language used to generate the label.
+     * @return A {@link Pair} object containing both the label and the value (Pair<label, value>).
+     */
+    public Pair<String, String> parseMetadata(Context context, Item item, String metadataField, String lang) {
+        // Retrieve the corresponding parser for the given metadata field.
+        MailMetadataParser parser = getParserForMetadataField(metadataField);
+        if (parser == null) {
+            logger.warn("Could not retrieve metadata parser for metadata field: [" + metadataField + "]");
+            return null;
+        }
+
+        // Get the parsed label using the provided language.
+        String label = parser.getLabel(lang) != null ? parser.getLabel(lang) : metadataField;
+        // Get the parsed value.
+        String value = getValue(context, item, parser, metadataField);
+
+        return Pair.of(label, value);
+    }
+
+    /**
+     * Parse a list of given metadata for the given item into a map of labels and values.
+     * 
+     * @param context The current DSpace context.
+     * @param item The item to extract values from.
+     * @param metadataFields The fields to extract from the item.
+     * @param lang The language used to generate the labels.
+     * @return A map containing all the labels and values for the given metadata fields.
+     * The map is of type {@link LinkedHashMap} which allows to keep the order of the metadataFields list.
+     */
+    public LinkedHashMap<String, String> parseMetadata(
+        Context context, Item item, List<String> metadataFields, String lang
+    ) {
+        return metadataFields.stream()
+            .map(field -> parseMetadata(context, item, field, lang))
+            .filter(pair -> pair != null)
```

Comment:
use `.filter(Objects::nonNull)` here. A good IDE would give you the suggestion :-)